File: python/questions/minCost.py
from typing import List
import heapq

class Solution:
    def minCost(self, n: int, edges: List[List[int]]) -> int:
        sdc = [[] for _ in range(n)]
        for e in edges:
            s, d, w = e[0], e[1], e[2]
            sdc[s].append([w, d])
            sdc[d].append([w*2, s])
        dmw = [float('inf') for _ in range(n)]
        hq = []
        dmw[0] = 0
        heapq.heappush(hq, [0, 0])
        while len(hq):
            (cw, cs) = heapq.heappop(hq)
            if cw > dmw[cs]:
                continue
            if cs == n-1:
                return cw
            for e in sdc[cs]:
                nw = cw + e[0]
                if nw < dmw[e[1]]:
                    dmw[e[1]] = nw
                    heapq.heappush(hq, [nw, e[1]])
        return -1

File: cpp/questions/minCost.cpp
#include <vector>
#include <queue>
#include <limits>

using namespace std;

class Solution {
public:
    int minCost(int n, vector<vector<int>>& edges) {
        std::vector<std::vector<std::pair<int, int>>> sdc(n);    
        for (auto& e: edges) {
            int s = e[0];
            int d = e[1];
            int w = e[2];
            sdc[s].push_back({d, w});
            sdc[d].push_back({s, w*2});
        }
        std::vector<int> gmc(n, std::numeric_limits<int>::max());
        std::priority_queue<std::pair<int, int>, std::vector<std::pair<int, int>>, std::greater<>> pq;
        gmc[0] = 0;
        pq.push({0, 0});
        while (!pq.empty()) {
            auto [w, d] = pq.top();
            pq.pop();
            if (w > gmc[d]) {
                continue;
            }
            if (d == n-1) {
                return w;
            }
            for (auto& [nd, aw]: sdc[d]) {
                int nw = w + aw;
                if (nw < gmc[nd]) {
                    gmc[nd] = nw;
                    pq.push({nw, nd});
                }
            }
        }
        return -1;
    }
};
